[Swift] Use Platform.h feature defines as compilation conditions

emw-apple opened the pull request:
#### 168c0aee29a2612fd6f7e71b66eba9a2f99171d5
<pre>
[Swift] Use Platform.h feature defines as compilation conditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=288691">https://bugs.webkit.org/show_bug.cgi?id=288691</a>
<a href="https://rdar.apple.com/problem/145717488">rdar://problem/145717488</a>

Reviewed by Alexey Proskuryakov.

Swift can and does import C/C++ headers from WTF, however, it imports
macro definitions as constant values. Unlike C, swift&apos;s #if expressions
are parsed separately from module imports, and cannot be influenced by
imported code. The way to influence compilation conditions is with -D
arguments, controlled by the SWIFT_ACTIVE_COMPILATION_CONDITIONS build
setting in Xcode.

The result of this is that we&apos;ve accumulated a handful of redeclarations
of out PlatformEnable.h flags in xcconfigs, which have to be carefully
maintained so that swift compilation matches clang. This change attempts
to fix this duplication and reestablish Platform.h as the source of
truth for WebKit&apos;s compile-time feature toggles.

Add a script to WTF that runs clang&apos;s preprocessor to generate an
arguments list of enabled features from &lt;wtf/Platform.h&gt;. This is
similar to what WebCore does in DerivedSources.make, however, it extends
WebCore&apos;s implementation by being able to solve simple indirection (e.g.
`#define ENABLE_FOO ENABLE_BAR` will look up the value of ENABLE_BAR).

Change the default OTHER_SWIFT_FLAGS xcconfig variable to include this
arguments list as a response file. Swift and Xcode record this as an
input dependency, so incremental changes to Platform.h and the arguments
list will invalidate swift code as well.

Replace some use of SWIFT_ACTIVE_COMPILATION_CONDITIONS with the
corresponding platform flags. I&apos;ve left WK_ENABLE_SWIFTUI in place,
because that build setting influences other build settings regarding
SwiftUI project bringup.

Swift&apos;s compilation conditions have the same semantics as our ENABLE() /
HAVE() macros, where an undeclared macro is implicitly undefined, so
there is no need to come up with a way to make a record of
deliberately-not-defined features.

* Configurations/CommonBase.xcconfig:
* Configurations/Sanitizers.xcconfig: As a drive-by fix, remove the &quot;=1&quot;
  to silence &quot;Conditional compilation flags do not have values&quot; warning.
* Source/WTF/Scripts/generate-platform-args: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: We need to generate one
  response file per compiler target. To do this while still declaring
  output dependencies, implement `generate-platform-args` as a build
  rule which is run once per architecture during the &quot;Compile Sources&quot;
  phase.
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
  WebKit_Private doesn&apos;t import successfully on macOS Sonoma, so use a
  compiler guard to only perform the import on newer releases. As a
  drive-by fix, remove a compiler guard that is no longer needed.

Canonical link: <a href="https://commits.webkit.org/291648@main">https://commits.webkit.org/291648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ce5e6ec95a23ee90e7c4bb8d6b6ae98fceab218

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93581 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/13152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2886 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/44107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21597 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/98584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/44107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96583 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/10043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84623 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/98584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/9725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/43421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/86291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/80000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/2309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100617 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92247 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/20633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/20885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/80560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79831 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19848 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/13781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/20617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/25795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114897 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/20304 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/23764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/22045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,43 @@
+swift_args = open(os.environ['SCRIPT_OUTPUT_FILE_0'], mode='w')
+
+feature_and_platform_defines = subprocess.check_output(
+    ('clang', '-std=c++2b', '-x', 'c++', '-target', target_triple,
```
Will it automatically know to use the correct toolchain, or do you need `xcrun` here?

Also, it's unfortunate to add another place where `-std` is defined, as this increases the risk of getting out of sync.

emw-apple replied:
> Will it automatically know to use the correct toolchain, or do you need xcrun here?

I think it relies on clang being `/usr/bin/clang`, to call into the xcselect library. I can throw a `xcrun` in here for good measure.

> Also, it's unfortunate to add another place where -std is defined, as this increases the risk of getting out of sync.

Good point. I bet we can pick up the relevant language build setting from here to avoid redefinition.

aproskuryakov reviewed (commented):
Beautiful and scary. Detailed review probably needs to wait until the bots are green, just a couple comments now.

Does this work with features enabled/disabled via `build-webkit` arguments?

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,43 @@
+    elif value in defined:
+        defined[name] = defined[value]
+
+for name, value in defined.items():
```
Do we need to sort this to avoid spurious rebuilds when order changes? I think that `clang` sorts, but wanted to ask to double check.

emw-apple replied:
I think that if `Platform.h` changes, this script's output is invalidated automatically, and the whole graph of dependencies underneath WTF rebuilds. So it doesn't really matter what we do here because it's always rebuilding in response to something else. I'll double check, though, if we can avoid recompiling all Swift in response to e.g. a Platform.h change that only affects other platforms, that's a nice win.

aproskuryakov replied:
There are two aspects here:
- whether we rebuild on incremental builds at all, determined by dependencies;
- whether compiler cache helps in clean builds.

Maybe (likely) sorting already happens for us, but we certainly don't want random order for the options in the resp file.

emw-apple replied:
CAS-based caching happens inside the compiler, whereas this is purely build system logic that influences compiler arguments. So unless the CAS is invalidated by the order of `-D` arguments, it shouldn't matter either way. That said…

> Maybe (likely) sorting already happens for us, but we certainly don't want random order for the options in the resp file.

I checked and you're right, we do get the definitions in sorted order. This made me realize that the logic to resolve simple indirections (i.e. `#define FOO BAR`) isn't good enough, because we don't receive items in definition order. Latest revision does the resolution as a separate step.

>  If we can avoid recompiling all Swift in response to e.g. a Platform.h change that only affects other platforms, that's a nice win.

Turns out that we _can_ avoid needless recompilation by only writing to the file when its contents will actually change. Latest revision does this.

emw-apple replied:
I played with this some more and it looks like the Swift driver is what's avoiding recompilation here, presumably because it keeps track of what compiler arguments it's invoked with. So going back to rewriting the .resp file every time Platform.h changes should be fine.

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,43 @@
+import re
+
+framework_search_paths = shlex.split('{BUILT_PRODUCTS_DIR} {FRAMEWORK_SEARCH_PATHS} '
+                                     '{SYSTEM_FRAMEWORK_SEARCH_PATHS}'.format_map(os.environ))
```
I'm a bit surprised that user and system search paths can be squashed together.

aproskuryakov commented on Source/WTF/WTF.xcodeproj/project.pbxproj:
```diff
@@ -972,6 +973,20 @@
+			dependencyFile = "$(DERIVED_FILES_DIR)/$(INPUT_FILE_BASE).$(arch).d";
+			filePatterns = "*/generate-platform-args";
+			fileType = pattern.proxy;
+			inputFiles = (
+			);
```
This isn't needed because there's a `dependencyFile`, correct?

emw-apple replied:
Exactly. The discovered dependency file will list wtf/Platform.h and all the headers transitively included.

aproskuryakov commented on Source/WTF/Scripts/generate-platform-args:
```diff
@@ -0,0 +1,43 @@
+for name, value in defined.items():
+    if name.startswith('HAVE_') or name.startswith('USE_') or name.startswith('ENABLE_') or \
+            name.startswith('WTF_PLATFORM') and value == 1:
+        swift_args.write(f'-D{name}\n')
```
I wonder if we want to print it out as a diagnostic for future investigations.

emw-apple replied:
That's a good idea—Xcode does this when it generates a response file, so we might as well do the same.

emw-apple committed:
[Swift] Use Platform.h feature defines as compilation conditions
https://bugs.webkit.org/show_bug.cgi?id=288691
rdar://problem/145717488

Reviewed by Alexey Proskuryakov.

Swift can and does import C/C++ headers from WTF, however, it imports
macro definitions as constant values. Unlike C, swift's #if expressions
are parsed separately from module imports, and cannot be influenced by
imported code. The way to influence compilation conditions is with -D
arguments, controlled by the SWIFT_ACTIVE_COMPILATION_CONDITIONS build
setting in Xcode.

The result of this is that we've accumulated a handful of redeclarations
of out PlatformEnable.h flags in xcconfigs, which have to be carefully
maintained so that swift compilation matches clang. This change attempts
to fix this duplication and reestablish Platform.h as the source of
truth for WebKit's compile-time feature toggles.

Add a script to WTF that runs clang's preprocessor to generate an
arguments list of enabled features from <wtf/Platform.h>. This is
similar to what WebCore does in DerivedSources.make, however, it extends
WebCore's implementation by being able to solve simple indirection (e.g.
`#define ENABLE_FOO ENABLE_BAR` will look up the value of ENABLE_BAR).

Change the default OTHER_SWIFT_FLAGS xcconfig variable to include this
arguments list as a response file. Swift and Xcode record this as an
input dependency, so incremental changes to Platform.h and the arguments
list will invalidate swift code as well.

Replace some use of SWIFT_ACTIVE_COMPILATION_CONDITIONS with the
corresponding platform flags. I've left WK_ENABLE_SWIFTUI in place,
because that build setting influences other build settings regarding
SwiftUI project bringup.

Swift's compilation conditions have the same semantics as our ENABLE() /
HAVE() macros, where an undeclared macro is implicitly undefined, so
there is no need to come up with a way to make a record of
deliberately-not-defined features.

* Configurations/CommonBase.xcconfig:
* Configurations/Sanitizers.xcconfig: As a drive-by fix, remove the "=1"
  to silence "Conditional compilation flags do not have values" warning.
* Source/WTF/Scripts/generate-platform-args: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj: We need to generate one
  response file per compiler target. To do this while still declaring
  output dependencies, implement `generate-platform-args` as a build
  rule which is run once per architecture during the "Compile Sources"
  phase.
* Source/WebKit/Configurations/Base.xcconfig:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
  WebKit_Private doesn't import successfully on macOS Sonoma, so use a
  compiler guard to only perform the import on newer releases. As a
  drive-by fix, remove a compiler guard that is no longer needed.

Canonical link: https://commits.webkit.org/291648@main
